lonestar/analytics/cpu/random-walk/RandomWalk.cpp

```diff
@@ -0,0 +1,631 @@
+
+    double lower_bound = 1.0f;
+
+    lower_bound = (lower_bound < prob_forward) ? lower_bound : prob_forward;
+    lower_bound = (lower_bound < prob_backward) ? lower_bound : prob_backward;
+
+    katana::do_all(
+        katana::iterate(graph),
+        [&](GNode n) {
+          std::uniform_real_distribution<double>* dist =
+              *distribution.getLocal();
+
+          katana::gstl::Vector<uint32_t> walk;
+          walk.push_back(n);
+
+          for (uint32_t walk_iter = 1; walk_iter <= walkLength; walk_iter++) {
+            if (walk_iter == 1) {
+              //random value between 0 and 1
+              double prob = (*dist)(*generator.getLocal());
+
+              //Assumption: All edges have weight 1
+              Graph::Node nbr;
+              FindSampleNeighbor(graph, n, degree, prob, &nbr);
+
+              walk.push_back(std::move(nbr));
+
+            } else {
+              uint32_t curr = walk[walk_iter - 1];
+              uint32_t prev = walk[walk_iter - 2];
+
+              //acceptance-rejection sampling
+              while (true) {
+                //sample x
+                double prob = (*dist)(*generator.getLocal());
+
+                Graph::Node nbr;
+                FindSampleNeighbor(graph, curr, degree, prob, &nbr);
+
+                //sample y
+                double y = (*dist)(*generator.getLocal());
+                y = y * upper_bound;
+
+                if (y <= lower_bound) {
+                  //accept this sample
+                  walk.push_back(std::move(nbr));
+                  break;
+                } else {
+                  //compute transition probability
+                  double alpha;
+
+                  //check if nbr is same as the previous node on this walk
+                  if (nbr == prev) {
+                    alpha = prob_backward;
+                  }  //check if nbr is also a neighbor of the previous node on this walk
+                  else if (
+                      katana::FindEdgeSortedByDest(graph, prev, nbr) !=
+                      graph.edge_end(prev)) {
+                    alpha = 1.0f;
+                  } else {
+                    alpha = prob_forward;
+                  }
+
+                  if (alpha >= y) {
+                    //accept y
+                    walk.push_back(std::move(nbr));
+                    break;
+                  }
+                }
+              }
+
+            }  //end if-else loop
+          }
+
+          num += 1;
+          (*walks).push(std::move(walk));
+        },
+        katana::steal(), katana::loopname("loop"));
+
+    katana::gPrint("num: ", num.reduce(), "\n");
+
+    for (uint32_t i = 0; i < distribution.size(); i++) {
+      delete (*distribution.getRemote(i));
+    }
+  }
+
+  void RandomWalks(
+      const Graph& graph,
+      katana::InsertBag<katana::gstl::Vector<uint32_t>>* walks,
+      const katana::LargeArray<uint64_t>& degree) {
+    uint32_t num_walks = 0;
+
+    while (num_walks < numWalks) {
```

Comment:
no, there is no reason as such. I can put it inside do_all too, but it will make it slightly complicated, that's all. Let me know what you think. thanks

Reply:
Why is it complicated?

By the way, I just realized: use `CHUNK_SIZE` of 1 for the `do_all` loops because each iteration is heavy-weight.